compiler/src/iree/compiler/ConstEval/JitGlobals.cpp

```diff
@@ -25,119 +28,297 @@ namespace mlir {
+    }
+
+    // Find immutable stores, early exiting if not supported.
+    // The consumers must come after rewrites of the producers above.
+    for (auto storeOp : funcOp.getOps<IREE::Util::GlobalStoreOp>()) {
+      auto globalOp = llvm::dyn_cast_or_null<IREE::Util::GlobalOp>(
+          sourceSymbolTable.lookup(storeOp.getGlobalAttr().getAttr()));
+      if (!globalOp || globalOp.getIsMutable()) {
+        emitWarning(storeOp.getLoc()) << "skipping consteval initializer: stor "
+                                         "to mutable globals not supported";
+        return failure();
+      }
+      Type t = storeOp.getValue().getType();
+      if (!supportedFeatures.isSupportedAbiType(t)) {
+        emitWarning(funcOp.getLoc())
+            << "skipping consteval initializer: unsupported type for current "
+               "jit configuration: "
+            << t;
+        return failure();
+      }
+
+      returns.push_back(storeOp.getValue());
+      returnTypes.push_back(t);
+      eraseOps.push_back(storeOp);
+      desc.resultBindings.emplace_back(globalOp);
+    }
+
+    // Cleanup.
+    for (auto *op : eraseOps) {
+      op->erase();
+    }
+
+    // Rewrite the terminator and the function type.
+    entryBlock->getTerminator()->erase();
+    OpBuilder termBuilder = OpBuilder::atBlockEnd(entryBlock);
+    termBuilder.create<func::ReturnOp>(funcOp.getLoc(), returns);
+    funcOp.setType(termBuilder.getFunctionType(argumentTypes, returnTypes));
+
+    jitFunctions.push_back(std::move(desc));
+    return success();
+  }
+
+  ModuleOp targetModuleOp;
   SymbolTable sourceSymbolTable;
   SymbolTable targetSymbolTable;
-  OpBuilder builder;
-  SmallVector<StringAttr> symbolImportWorklist;
-};
-
-// These options structs are not copy-constructable so we have to allocate them
-// shared.
-// TODO: See if we can make them copyable?
-struct CompileOptions {
-  BindingOptions bindingOptions;
-  InputDialectOptions inputOptions;
-  PreprocessingOptions preprocessingOptions;
-  HighLevelOptimizationOptions highLevelOptimizationOptions;
-  SchedulingOptions schedulingOptions;
-  IREE::HAL::TargetOptions executableOptions;
-  IREE::VM::TargetOptions targetOptions;
-  IREEVMPipelineHooks hooks;
+  llvm::SmallVector<JitFunctionDesc> jitFunctions;
+  const SupportedFeatures &supportedFeatures;
 };
 
 struct JitGlobalsPass : public JitGlobalsBase<JitGlobalsPass> {
-  JitGlobalsPass()
+  JitGlobalsPass(const IREE::HAL::TargetBackendRegistry &targetRegistry)
       : options(std::make_shared<CompileOptions>()),
         compilePipeline("builtin.module") {
-    // Invoke IREE compilation flow.
-    options->executableOptions.targets.push_back("vmvx");
-    options->targetOptions.f32Extension = true;
-    options->targetOptions.f64Extension = false; // not yet implemented
+    // Detect backend.
+    hasLLVMCPUBackend = targetRegistry.getTargetBackend("llvm-cpu") != nullptr;
+    if (clUseVMVX || !hasLLVMCPUBackend) {
```

Comment:
could make the flag a string `iree-consteval-jit-backend` - the vmvx target options here are the defaults and shouldn't be needed so the only thing changing is which string gets pushed back - could still query if the requested backend exists and fall back to hardcoded vmvx if not or something

Reply:
Mind if I take a pass for now? This was just for my own testing and I'm working on the followup that fixes this right.